gapis/replay/mapping_exporter.go

```diff
@@ -30,74 +29,91 @@ import (
 	"github.com/google/gapid/gapis/api/transform"
 	"github.com/google/gapid/gapis/memory"
 	"github.com/google/gapid/gapis/replay/builder"
+	"github.com/google/gapid/gapis/replay/value"
 	"github.com/google/gapid/gapis/service"
 )
 
+type mappingHandle struct {
+	traceValue    uint64
+	replayAddress value.Pointer
+	size          uint64
+	name          string
+}
+
 type MappingExporter struct {
-	mappings *map[uint64][]service.VulkanHandleMappingItem
-	thread   uint64
-	path     string
-	results  map[uint64]service.VulkanHandleMappingItem
-	mutex    sync.RWMutex
+	mappings       *map[uint64][]service.VulkanHandleMappingItem
+	thread         uint64
+	path           string
+	traceValues    []mappingHandle
+	notificationID uint64
 }
 
 func NewMappingExporter(ctx context.Context, mappings *map[uint64][]service.VulkanHandleMappingItem) *MappingExporter {
 	return &MappingExporter{
-		mappings: mappings,
-		thread:   0,
-		path:     "",
-		results:  make(map[uint64]service.VulkanHandleMappingItem),
-		mutex:    sync.RWMutex{},
+		mappings:       mappings,
+		thread:         0,
+		path:           "",
+		traceValues:    make([]mappingHandle, 0, 0),
+		notificationID: 0,
 	}
 }
 
 func NewMappingExporterWithPrint(ctx context.Context, path string) *MappingExporter {
 	mapping := make(map[uint64][]service.VulkanHandleMappingItem)
 	return &MappingExporter{
-		mappings: &mapping,
-		thread:   0,
-		path:     path,
-		results:  make(map[uint64]service.VulkanHandleMappingItem),
-		mutex:    sync.RWMutex{},
+		mappings:       &mapping,
+		thread:         0,
+		path:           path,
+		traceValues:    make([]mappingHandle, 0, 0),
+		notificationID: 0,
 	}
 }
 
 func (m *MappingExporter) processNotification(ctx context.Context, s *api.GlobalState, n gapir.Notification) {
-	notificationData := n.GetData()
-	notificationID := n.GetId()
-	mappingData := notificationData.GetData()
-
-	m.mutex.Lock()
-
-	result, ok := m.results[notificationID]
-	if !ok {
-		log.I(ctx, "Invalid notificationID %d", notificationID)
+	if m.notificationID != n.GetId() {
+		log.I(ctx, "Invalid notificationID %d", m.notificationID)
 		return
 	}
 
+	notificationData := n.GetData()
+	mappingData := notificationData.GetData()
+
 	byteOrder := s.MemoryLayout.GetEndian()
 	r := endian.Reader(bytes.NewReader(mappingData), byteOrder)
-	replayValue := r.Uint64()
-	result.ReplayValue = replayValue
 
-	if _, ok := (*m.mappings)[replayValue]; !ok {
-		(*m.mappings)[replayValue] = make([]service.VulkanHandleMappingItem, 0, 0)
-	}
+	for _, handle := range m.traceValues {
+		var replayValue uint64
+		switch handle.size {
+		case 1:
+			replayValue = uint64(r.Uint8())
+		case 2:
+			replayValue = uint64(r.Uint16())
+		case 4:
+			replayValue = uint64(r.Uint32())
+		case 8:
+			replayValue = r.Uint64()
+		}
```

Comment:
Add a `default:` and treat it as an error (not sure what is best: early return? panic?)

Reply:
Given that this is something that should never happen, I think it's better to have an early crash through a fatal log, i.e. `log.F(ctx, true, ...)`. Otherwise, we just get an error in the log but the program keeps on running with replayValue set to its default (zero), and this can lead to other issues/crash much further down, which may be harder to link to an invalid size here. With an early crash, we know directly where things started to go wrong.

Reply:
Great, now last nit: upon error, it would be good to print the `handle.size` alongside the name. I approve the PR and let you add this last thing before merging. Thanks!

Reply:
Brilliant, thanks!